From 52a880b36efd8d23295a96ec5c1c78409a47689a Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Tue, 14 Jan 2020 20:46:29 +0200
Subject: [PATCH] [SW-8228] arcboot: add efuse support

add support to read the efuses, code based on uboot
implimentation. added ORing between bank0 and bank1
valuse.

Change-Id: I4bb68f3a67bd4756c556ff129750575447106c30
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/common/efuse.c b/common/efuse.c
new file mode 100644
index 0000000..ec3ce46
--- /dev/null
+++ b/common/efuse.c
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2016-2019 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#include <unistd.h>
+#include "arcboot.h"
+
+#define FUSE_TIMEOUT_US		100000
+#define FUSE_DELAY_US		20
+#define FUSE_RETRIES		(FUSE_TIMEOUT_US / FUSE_DELAY_US)
+#define FUSE_WORDS_PER_BANK	64
+
+/*
+ * fuse box is devided into half for redundancy
+ * each half is bank of FUSE_WORDS_PER_BANK words
+ * NOTE: bank1 is a complete duplication of bank0
+ */
+int fuse_read(u32 word, u32 *val)
+{
+	int i, offset, delay;
+
+	if (word > 64) {
+		return -1;
+	}
+
+	regmap_writel(1, mmPSOC_EFUSE_READ);
+
+	for (i=0; i < FUSE_RETRIES; i++)
+		if (regmap_readl(mmPSOC_EFUSE_READ))
+			for(delay=FUSE_DELAY_US; delay > 0; delay--) asm("nop");
+		else
+			break;
+
+	if (i == FUSE_RETRIES) {
+		return -2;
+	}
+
+	// read value from bank0
+	offset = (word * sizeof(u32));
+	*val = regmap_readl(mmPSOC_EFUSE_DATA_0 + offset);
+
+	// read value from bank1 and OR with value from bank0
+	offset = ((FUSE_WORDS_PER_BANK + word) * sizeof(u32));
+	*val |= regmap_readl(mmPSOC_EFUSE_DATA_0 + offset);
+
+	return 0;
+}
diff --git a/common/efuse.h b/common/efuse.h
new file mode 100644
index 0000000..ae073e5
--- /dev/null
+++ b/common/efuse.h
@@ -0,0 +1,13 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef _EFUSE_H_
+#define _EFUSE_H_
+
+int fuse_read(u32 word, u32 *val);
+
+#endif /* _EFUSE_H_ */
\ No newline at end of file
